citrixitm/resource_citrixitm_dns_app.go

```diff
@@ -56,57 +58,60 @@ func resourceCitrixITMDnsApp() *schema.Resource {
 	}
 }
 
-func resourceCitrixITMDnsAppCreate(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*itm.Client)
+func resourceCitrixITMDnsAppCreate(d *schema.ResourceData, m interface{}) error {
+	client := m.(*itm.Client)
 	opts := itm.NewDNSAppOpts(
 		d.Get("name").(string),
 		d.Get("description").(string),
 		d.Get("fallback_cname").(string),
 		d.Get("app_data").(string),
 	)
-	log.Printf("[DEBUG] DNS app create options: %#v", opts)
+	log.Printf("[DEBUG] %s create options: %#v", resourceName, opts)
 	app, err := client.DNSApps.Create(&opts, true)
 	if err != nil {
 		return nil
 	}
-	d.SetId(fmt.Sprintf("%d", app.Id))
-	return read(app.Id, client, d)
-}
-
-type ProcessAppFunc func(id int, c *itm.Client, d *schema.ResourceData) error
-
-func withExistingResource(f ProcessAppFunc) func(*schema.ResourceData, interface{}) error {
-	return func(d *schema.ResourceData, m interface{}) error {
-		id, err := strconv.Atoi(d.Id())
-		if err != nil {
-			return fmt.Errorf("Invalid app id: %s", d.Id())
-		}
-		return f(id, m.(*itm.Client), d)
-	}
+	d.SetId(strconv.Itoa(app.Id))
+	log.Printf("[INFO] %s id: %s", resourceName, d.Id())
+	return resourceCitrixITMDnsAppRead(d, m)
 }
 
-func read(id int, c *itm.Client, d *schema.ResourceData) error {
-	app, err := c.DNSApps.Get(id)
-	log.Printf("[DEBUG] Inside read; app: %#v", app)
+func resourceCitrixITMDnsAppRead(d *schema.ResourceData, m interface{}) error {
+	id, err := strconv.Atoi(d.Id())
 	if err != nil {
-		return fmt.Errorf("Error retrieving app: %s", err)
+		return fmt.Errorf("Error converting app id (%s) to an integer: %s", d.Id(), err)
 	}
-	if app.Enabled {
-		d.Set("name", app.Name)
-		d.Set("description", app.Description)
-		d.Set("fallback_cname", app.FallbackCname)
-		d.Set("fallback_ttl", app.FallbackTtl)
-		d.Set("app_data", app.AppData)
-		d.Set("cname", app.AppCname)
-		d.Set("version", app.Version)
-	} else {
-		log.Printf("The app is disabled. This likely means that it was deleted outside of Terraform. 'terraform apply' will recreate the app if you approve. If you don't wish Terraform to continue prompting about it, then you may want to remove its configuration.")
+	client := m.(*itm.Client)
+	app, err := client.DNSApps.Get(id)
+	if err != nil {
+		// There was a problem retrieving the app
+		// Set the resource ID to "" to indicate that the resource is not present.
+		log.Printf("[WARN] %s id (%s) not found", resourceName, d.Id())
 		d.SetId("")
+	} else {
+		if app.Enabled {
+			d.Set("name", app.Name)
+			d.Set("description", app.Description)
+			d.Set("fallback_cname", app.FallbackCname)
+			d.Set("fallback_ttl", app.FallbackTtl)
+			d.Set("app_data", app.AppData)
+			d.Set("cname", app.AppCname)
+			d.Set("version", app.Version)
```

Comment:
shouldn't there be a `d.SetId(id)` here?

Reply:
No, because it's already assumed to be set. That's how we were able to make the API request in the first place. AFAIK a resource ID should never be modified without creating a brand new resource, which is triggered if you call `d.SetId("")` and then expect Terraform to go back through the "create" function in order to set the new ID value.